apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/plugin/bytebuddy/CacheableTransformerDecorator.java

```diff
@@ -0,0 +1,195 @@
+    private static final ILog logger = LogManager.getLogger(CacheableTransformerDecorator.class);
+
+    private final ClassCacheMode cacheMode;
+    private ClassCacheResolver cacheResolver;
+
+    public CacheableTransformerDecorator(ClassCacheMode cacheMode) throws IOException {
+        this.cacheMode = cacheMode;
+        initClassCache();
+    }
+
+    private void initClassCache() throws IOException {
+        if (this.cacheMode.equals(ClassCacheMode.FILE)) {
+            String cacheDirBase = null;
+            try {
+                cacheDirBase = AgentPackagePath.getPath() + "/class-cache";
+            } catch (AgentPackageNotFoundException e) {
+                throw new IOException("Can't find the root path for creating /class-cache folder.");
+            }
+            File cacheDir = new File(cacheDirBase + "/class-cache-" + RandomString.make());
+            if (!cacheDir.exists()) {
+                cacheDir.mkdirs();
+            }
+            if (!cacheDir.exists()) {
+                throw new IOException("Create class cache dir failure");
+            }
+
+            cacheResolver = new FileCacheResolver(cacheDir);
+        } else {
+            cacheResolver = new MemoryCacheResolver();
+        }
+    }
+
+    @Override
+    public ResettableClassFileTransformer decorate(ResettableClassFileTransformer classFileTransformer) {
+        return new ResettableClassFileTransformer.WithDelegation(classFileTransformer) {
+
+            @Override
+            public byte[] transform(ClassLoader loader, String className, Class<?> classBeingRedefined, ProtectionDomain protectionDomain, byte[] classfileBuffer) throws IllegalClassFormatException {
+                // load from cache
+                byte[] classCache = cacheResolver.getClassCache(loader, className);
+                if (classCache != null) {
+                    return classCache;
+                }
+
+                //transform class
+                classfileBuffer = classFileTransformer.transform(loader, className, classBeingRedefined, protectionDomain, classfileBuffer);
+
+                // save to cache
+                if (classfileBuffer != null) {
+                    cacheResolver.putClassCache(loader, className, classfileBuffer);
+                }
+
+                return classfileBuffer;
+            }
+        };
+    }
+
+    private static String getClassLoaderHash(ClassLoader loader) {
+        String classloader;
+        if (loader != null) {
+            classloader = Integer.toHexString(loader.hashCode());
+        } else {
+            //classloader is null for BootstrapClassLoader
+            classloader = "00000000";
+        }
+        return classloader;
+    }
+
+    interface ClassCacheResolver {
+
+        byte[] getClassCache(ClassLoader loader, String className);
+
+        void putClassCache(ClassLoader loader, String className, byte[] classfileBuffer);
+    }
+
+    static class MemoryCacheResolver implements ClassCacheResolver {
```

Comment:
This cache doesn't seem to limit whether or not this puts a strain on the memory

Reply:
All relative things here have been removed as new kernel implementation in 9.x, which has better support for transfer. Don't worry about this anymore, and pick 9.x latest release.

Reply:
ok. thx